car-bridge/src/main.rs

```diff
@@ -29,58 +41,198 @@ async fn main() -> Result<(), Box<dyn Error>> {
-    info!("Subscribing to topic '{c2d_topic}'.");
-    client.subscribe(c2d_topic.clone(), QOS_2).await?;
+    let publish_handle = {
+        // Detach sending the responses from handling the messages to avoid
+        // backpressure and disconnect the client gracefully.
 
-    // Publish a phony message.
-    client.publish(Message::new(c2d_topic, "Example payload", QOS_2)).await?;
+        let cancellation_token = cancellation_token.child_token();
+        spawn(async move {
+            let client = Arc::new(client);
 
-    let cancellation_token = ctrl_c_cancellation();
+            loop {
+                select! {
+                    message = response_receiver.recv() => {
+                        let Some((topic, message)) = message else {
+                            warn!("Response receiver stopped, no more messages will be published.");
+                            break;
+                        };
+
+                        let client = Arc::clone(&client);
+
+                        spawn(async move {
+                            if let Err(e) = client.publish(topic, message).await {
+                                debug!("Error when publishing message: '{:?}'.", e);
+                            }
+                        });
+                    }
+                    _ = cancellation_token.cancelled() => {
+                        debug!("Shutting down the publisher loop.");
+                        break;
+                    }
+                }
+            }
+        })
+    };
 
     loop {
         select! {
-            message = message_stream.next() => {
+            message = messages.next() => {
                 let Some(message) = message else {
                     break;
                 };
-                if let Some(message) = message {
-                    info!("(R) {message:?}");
-                }
-                else {
-                    info!("Connection temporarily lost.");
 
-                    while let Err(err) = client.reconnect().await {
-                        info!("Trying to reconnect: {}.", err);
-                        sleep(Duration::from_secs(5)).await;
-                    }
-                }
+                let mut chariott = chariott.clone();
+                let response_sender = response_sender.clone();
+
+                spawn(async move {
+                    handle_message(&mut chariott, response_sender, message).await;
+                });
             }
             _ = cancellation_token.cancelled() => {
+                debug!("Shutting down the subscriber loop.");
                 break;
             }
         }
     }
 
-    info!("Disconnecting the client.");
-    client.disconnect(None).await?;
+    publish_handle.await?;
 
     Ok(())
 }
+
+async fn handle_message(
+    chariott: &mut impl ChariottCommunication,
+    response_sender: Sender<(String, MessageBuilder)>,
+    message: MqttMessage,
+) {
+    async fn get_response(
```

Comment:
Curious why you use a local function for something that's just used/called once?

Reply:
Yes, the idea is that for every fallible operation within `get_response` the action that we take is the same: we wrap the error message as an error response. For errors that happen outside of the `get_response` we either don't have the correlation data yet (and hence do not know where to return an error), or get an error that we can not recover from (which we log instead). The local function allows  us to handle all "error messages" once and in a unified way.

Reply:
See [suggestion on inlining this](https://github.com/eclipse/chariott/pull/52/files#r1031475149).